### DISCUSSION TITLE
cannot save record - "field is read only"

### DISCUSSION THREAD

**Discussion by teafmartins** [serenity-is/Serenity#5312] [Help]
Hello,

I have background sevices that update discount field in order database table.
Field "Discount" in Row.cs is defined as "Updatable(false)", but when I save an order tha has changed discount field I get message "discount field is read only".

        [DisplayName("Discount"), DefaultValue(0), AlignRight, DisplayFormat("#,##0.00"), Updatable(false)]
        public Single? Discount
        {
            get { return Fields.Discount[this]; }
            set { Fields.Discount[this] = value; }
        }

![image](https://user-images.githubusercontent.com/17210232/102729267-f2218280-4327-11eb-9273-d3d248822ba8.png)

Is this a bug? shoudn't serenity bypass update of "discount" field?

Thanks for you help.

**Comment by VictorTomaili** [comment]
I transfer this question to discussion. 
You need clear assignment before save and before update in your order repository.

**Reply by teafmartins** [reply]
@VictorTomaili , sorry again, but what do you mean by "clear assignment"?

**Reply by teafmartins** [reply]
if I use SetFieldFlags(FieldFlags.Calculated) should do it, right?

**Comment by JohnRanger** [comment]
Hi @teafmartins ,

I don't completely follow your use case :-) What is the goal? Allow the discount value to be changed or not? If it should be changeable (be it in GUI or background service), then the field within xyzRow.cs should *not* be declared as updateable(false) IMHO - because it in fact *IS* updateable. 

If you want to update the discount field within the "standard" xyzRepository.cs onBeforeSave method, then you also cannot set the [readonly] attribute within xyzForm.cs (as otherwise you get the same alert). In such cases, set the field programmatically to "readonly" within the xyzDialog.cs by using:

```ts
Serenity.EditorUtils.setReadOnly(this.form.<discountFieldName>, true);
```

within afterLoadEntity method. Ten it will behave everywhere as a read-writeable field - except within the dialog.

Hope this helps.

With kind regards,

John

**Reply by teafmartins** [reply]
hi @JohnRanger 
No, I don't want discount field to be updated using GUI. 
I want serenity to **keep the value of discount field at the time of saving ** (which can be different from when I opened the dialog).
I have one service, **independent from serenity**, that updates directly my databse and that field.

Right now I'm defining the field with "SetFieldFlags(FieldFlags.Calculated)". This prevents discount field data to be overlapped when I save it from serenity GUI. Is this the better way to do it?

Thanks for your time,
Regards,
Tiago

**Reply by JohnRanger** [reply]
Hmm , ok - I understand now your use case. 

But then the readonly tag should be ok - as long as your underlying table does not have "not null" on the field - and also within your xyzRow.cs on the field you must not have the "not null" tag.

Because otherwise there will be a value required within the discount field when creating the record - either by Serenity validation (when you have the "not null" tag within xyzRow.cs - what was shown within your screenshot) - or you will get a SQL error bubbling up when trying to save a new record which then will be displayed within your app.

I would use the readonly tag on the field within xyzRow.cs.

With kind regards,

John
